defi/marketplace/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+mod data;
+
+use scrypto::prelude::*;
+
+use data::*;
+
+blueprint! {
+    struct Market {
+        order_count: i64,
+        currency: Address,
+        buy_orders: Vec<Order>,
+        sell_orders: Vec<Order>,
+        ticket_minter_badge: Vault,
+        market_prices: MarketPrices
+    }
+
+    impl Market {
+        pub fn open(currency: Address) -> Component {
+            let ticket_minter_badge = ResourceBuilder::new()
+                .metadata("name", "Order Ticket Minter Badge")
+                .new_badge_fixed(1);
+
+            Self {
+                order_count: 0,
+                currency: currency,
+                buy_orders: vec![],
+                sell_orders: vec![],
+                ticket_minter_badge: Vault::with_bucket(ticket_minter_badge),
+                market_prices: MarketPrices::new()
+            }
+                .instantiate()
+        }
+
+        /// Yields a new token specifically for this order which can be used to withdraw
+        /// from it once it's filled.
+        fn order_ticket(&self, order_number: i64, token_address: Address) -> (Bucket, Address) {
+            let ticket_resource_def = ResourceBuilder::new()
+                .metadata("name", format!("Order Ticket #{}", order_number))
+                .metadata("symbol", format!("OT-{}", order_number))
+                .metadata("order_number", format!("{}", order_number))
+                .metadata("order_token_address", token_address.to_string())
+                .metadata("order_currency", self.base_currency())
+                .new_token_mutable(self.ticket_minter_badge.resource_def());
+
+            self.ticket_minter_badge.authorize(|badge|{
+                (ticket_resource_def.mint(1, badge), ticket_resource_def.address())
+            })
+        }
```

Comment:
With Scrypto v0.2.0, we can now use NFT to represent a ticket without creating too many resources.